### PR TITLE
qemu_v8.mk: Add CI test for Rust applications

### PR DESCRIPTION
**Pull request by DemesneGH**
Invoke optee_rust/ci/qemu-check.exp in target "check-rust" to
trigger CI test for Rust applications.
Run check progress:
$ make CFG_TEE_CORE_LOG_LEVEL=0 check-rust

Signed-off-by: Yuan Zhuang <zhuangyuan04@baidu.com>

### CODE REVIEW

**Comment by DemesneGH** [comment]
Related PR:
https://github.com/OP-TEE/manifest/pull/202

**Review by jforissier — COMMENTED** [review]
Hi @DemesneGH, thanks for the patch. Please see my comments below.

One other thing, not strictly related to the CI script. The Rust examples take a long time to build on my laptop, around 25 seconds for each example. Is that expected? Since there are 16 of them that is a lot of time, and even when there is nothing to do each example takes several seconds. Why is each example compiling the same things again and again? uuid, proto, quote, syn... I don't know what I'm talking about ;-) but perhaps there is room for improvement, some sharing or caching perhaps? (what about [sccache](https://lib.rs/crates/sccache)?)
The build log mentions "[optimized] targets", could some time be spared by not optimizing?
Also the build is not parallel which is unfortunate. Perhaps the Buildroot makefile could be written differently to be more efficient (I know we have the same issue with the C examples but they build much more quickly).

I am raising this speed issue because I suspect the next step will be to add the `make check-rust` command as a new job to the `optee_os` [CI script](https://github.com/OP-TEE/optee_os/blob/master/.azure-pipelines.yml). For this to be reasonable, it would need to not significantly increase the overall CI time, which means 25-30 minutes max for the Rust check (see [here](https://dev.azure.com/OPTEE/optee_os/_build?view=runs).

Anyways, with my comments below addressed:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`
`Tested-by: Jerome Forissier <jerome@forissier.org>`

**Comment by mssun** [comment]
> The Rust examples take a long time to build on my laptop, around 25 seconds for each example.

Firstly, the Rust compiler is indeed slow compared with C compiler. => We cannot do anything with this.

Specifically, for this project, examples are self-contained and standalone crates, meaning that they don't share compilation objects. Also, for each TA, the Rust's standard library with OP-TEE target is also compiled. => We need to re-structure the organization of examples, e.g., put host app and TA in workspaces to share compilation artifacts. This may need some time to be improved.

Also, I think the default is the "release" build. Therefore, it may also take time to compiling and linking. => This can be solve by using debug build by default.

File: qemu_v8.mk [hunk 1]
```diff
@@ -429,6 +429,23 @@ check: $(CHECK_DEPS)
 			echo "== end of $$PWD/serial1.log:"; \
 		fi; false)
 
+check-rust: $(CHECK_DEPS)
```
[the remark this answers is not in the document]
Comment (DemesneGH) [reply qemu_v8.mk:434]: Now we can run build and check progress by:
```
$ make CFG_TEE_CORE_LOG_LEVEL=0 OPTEE_RUST_ENABLE=y check-rust
```

**Comment by jforissier** [comment]
> => We need to re-structure the organization of examples, e.g., put host app and TA in workspaces to share compilation artifacts. This may need some time to be improved.

Yes that would be probably be good, nothing urgent though.

> Also, I think the default is the "release" build. Therefore, it may also take time to compiling and linking. => This can be solve by using debug build by default.

Yes. Debug build takes 8 minutes, release build takes between 16 and 19 minutes. So it would be good to use debug by default, that is a very easy thing to do.

If we add parallel build (Rust example Makefile + build.git diff below), build time further drops to 6 minutes.
```
$ cat optee_rust/examples/Makefile
EXAMPLE_LIST := $(filter-out Makefile,$(wildcard *))

all: $(EXAMPLE_LIST)

define gen-target
.PHONY: $(1)
$(1):
	$$(MAKE) -C $(1)
endef

$(foreach d,$(EXAMPLE_LIST),$(eval $(call gen-target,$(d))))
```
```diff
diff --git a/br-ext/package/optee_rust_examples_ext/optee_rust_examples_ext.mk b/br-ext/package/optee_rust_examples_ext/optee_rust_examples_ext.mk
index 09918f0..a1fabbe 100644
--- a/br-ext/package/optee_rust_examples_ext/optee_rust_examples_ext.mk
+++ b/br-ext/package/optee_rust_examples_ext/optee_rust_examples_ext.mk
@@ -20,9 +20,7 @@ export VENDOR = qemu_v8.mk
 export OPTEE_OS_INCLUDE = $(OPTEE_DIR)/optee_os/out/arm/export-ta_arm64/include

 define OPTEE_RUST_EXAMPLES_EXT_BUILD_CMDS
-       @$(foreach f,$(wildcard $(@D)/examples/*/Makefile), \
-               echo Building $f && \
-               $(MAKE) -C $(dir $f) &&) true
+               +$(MAKE) -C $(@D)/examples
 endef

 define OPTEE_RUST_EXAMPLES_EXT_INSTALL_TARGET_CMDS
```

**Review by jforissier — APPROVED** [review]
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Pull request merged**